### PR TITLE
feat: implementation of csidriver resource unit

### PR DESCRIPTION
**Pull request by leelavg**
- implemented create and delete reconcile actions on CSIDriver resource
- added rbac roles to manager to manager csidriver resource

Signed-off-by: Leela Venkaiah G <lgangava@redhat.com>

### CODE REVIEW

**Comment by leelavg** [comment]
- will add tests after reaching a consense on #17

File: controllers/topolvm_csi_driver.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	csiDriverResource := &storagev1.CSIDriver{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: TopolvmCSIDriverName}, csiDriverResource)
+
+	if err != nil {
+		// already deleted in previous reconcile
+		if errors.IsNotFound(err) {
+			r.Log.Info("csi driver deleted", "TopolvmCSIDriverName", csiDriverResource.Name)
+			return nil
+		}
+		r.Log.Error(err, "unable to retrieve topolvm csi driver resource", "TopolvmCSIDriverName", csiDriverResource.Name)
```
Comment (nbalacha) [inline controllers/topolvm_csi_driver.go:58]: "failed to retrieve"...

File: controllers/topolvm_csi_driver.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+	// if not deleted, initiate deletion
+	if csiDriverResource.GetDeletionTimestamp().IsZero() {
+		if err = r.Client.Delete(ctx, csiDriverResource); err != nil {
+			r.Log.Error(err, "unable to delete topolvm csi driver", "TopolvmCSIDriverName", csiDriverResource.Name)
+			return err
+		}
+	} else {
+		// set deletion in-progress for next reconcile to confirm deletion
+		return fmt.Errorf("topolvm csi driver %s is being uninstalled", csiDriverResource.Name)
```
Comment (nbalacha) [inline controllers/topolvm_csi_driver.go:70]: "topolvm csi driver %s is already marked for deletion"

**Review by nbalacha — CHANGES_REQUESTED** [review]
Some minor comments. Also, please split this into 2 commits - 1 for the code changes and 1 for the manifest changes after regenerating them.

**Comment by leelavg** [comment]
- tested against a live cluster by deploying, updating and deleting lvmcluster CR and observing reconciliation of csidriver resource unit
- reg addition of list & watch verbs, operator-sdk seems to be building up sharedinformercache using client-go's watch&list functions and so added those verbs as well

File: controllers/topolvm_csi_driver.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+func (c csiDriver) getName() string {
+	return driverName
+}
+
+//+kubebuilder:rbac:groups=storage.k8s.io,resources=csidrivers,verbs=get;create;delete;watch;list
+
+func (c csiDriver) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	csiDriverResource := getCSIDriverResource()
+	result, err := cutil.CreateOrUpdate(ctx, r.Client, csiDriverResource, func() error {
+		return nil
```
Comment (sp98) [inline controllers/topolvm_csi_driver.go:33]: why no `Mutatefn` implementation?  Is the lvmCluster not supposed to be updated once its created?

Comment (leelavg) [reply controllers/topolvm_csi_driver.go:33]: - yes, it's a clusterscoped resource with most of the fields as immutable and mutable fields (name & k8s managed) need not to be mutated

**Review by sp98 — APPROVED** [review]

**Review by jmolmo — APPROVED** [review]

File: controllers/lvmcluster_controller_test.go [hunk 4]
```diff
@@ -67,6 +79,13 @@ var _ = Describe("LVMCluster controller", func() {
 			Eventually(func() bool {
 				err := k8sClient.Delete(ctx, lvmClusterOut)
```
Comment (nbalacha) [inline controllers/lvmcluster_controller_test.go:80]: Change this to do the following:
Delete the lvmCluster. client.Delete should be successful
Eventually get(CSIDriver) should return errors.IsNotFound
Eventually get( LVMCluster) should return errors.IsNotFound

**Review by nbalacha — CHANGES_REQUESTED** [review]
Please refactor the delete test

**Review by nbalacha — APPROVED** [review]
/lgtm

**Pull request merged**